zerver/tests/test_message_fetch.py

```diff
@@ -1534,6 +1545,31 @@ def test_unauthenticated_get_messages_with_web_public(self) -> None:
             result, "Not logged in: API authentication or user session required", status_code=401
         )
 
+    def test_unauthenticated_get_messages_disabled_spectator_login(self) -> None:
+        """
+        An unauthenticated call to GET /json/messages without valid
+        parameters in the `streams:web-public` narrow returns a 401.
+        """
+        self.realm.enable_spectator_access = False
+        self.realm.save()
+        post_params: Dict[str, Union[int, str, bool]] = {
+            "anchor": 1,
+            "num_before": 1,
+            "num_after": 1,
+            "narrow": orjson.dumps(
+                [
+                    dict(operator="streams", operand="web-public"),
+                    dict(operator="stream", operand="Scotland"),
+                ]
+            ).decode(),
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+        self.assert_json_error(
+            result, "Not logged in: API authentication or user session required", status_code=401
+        )
+        self.realm.enable_spectator_access = True
+        self.realm.save()
```

Comment:
You don't need to reset any state at the end of tests -- the test system runs each test in a transaction and reverts the transaction at the end, so no database updates should stick.  (Though the way you did the `setUp` method might have resulted in a stale object hanging around).

Reply:
Hmm, the start of https://zulip.readthedocs.io/en/latest/testing/testing-with-django.html#setting-up-data-for-tests tries to explain this, but I think it doesn't actually spell out that this means you don't need to do anything to reset state at the end of tests?

Reply:
Okay, will remember.